src/stan/services/pathfinder/psis.hpp

```diff
@@ -0,0 +1,330 @@
+  } else {
+    return std::make_pair(x.eval(), k);
+  }
+}
+
+/**
+ * This function takes the last element as pivot, places the pivot element at
+ * its correct position in the sorted array, and places all values smaller than
+ * the pivot to the left of the pivot and all greater elements to the right of
+ * the pivot
+ * @param[in, out] arr The Array of doubles to be sorted
+ * @param[in, out] idx The index of the original positions of the elements of
+ * `arr`. This is also sorted to keep track of the original positions of the
+ * elements in `arr`.
+ * @param[in] low Starting index of the sort
+ * @param[in] high Ending index of the sort
+ */
+inline Eigen::Index quick_sort_partition(Eigen::Array<double, -1, 1>& arr,
+                                         Eigen::Array<Eigen::Index, -1, 1>& idx,
+                                         const Eigen::Index low,
+                                         const Eigen::Index high) {
+  const double pivot = arr.coeff(high);  // pivot
+  Eigen::Index i = (low - 1l);           // Index of smaller element
+  for (Eigen::Index j = low; j <= high - 1; ++j) {
+    // If current element is smaller than or
+    // equal to pivot
+    if (arr.coeff(j) <= pivot) {
+      ++i;  // increment index of smaller element
+      std::swap(arr.coeffRef(i), arr.coeffRef(j));
+      std::swap(idx.coeffRef(i), idx.coeffRef(j));
+    }
+  }
+  std::swap(arr.coeffRef(i + 1l), arr.coeffRef(high));
+  std::swap(idx.coeffRef(i + 1l), idx.coeffRef(high));
+  return (i + 1l);
+}
+
+/**
+ * Runs quick_sort optionally in parallel
+ * @tparam Parallel Whether to allow the algorithm to attempt to run in
+ * parallel.
+ * @param[in, out] arr The Array of doubles to be sorted
+ * @param[in, out] idx The index of the original positions of the elements of
+ * `arr`. This is also sorted to keep track of the original positions of the
+ * elements in `arr`.
+ * @param[in] low Starting index of the sort
+ * @param[in] high Ending index of the sort
+ */
+template <bool DoParallel = true>
+inline void quick_sort(Eigen::Array<double, -1, 1>& arr,
+                       Eigen::Array<Eigen::Index, -1, 1>& idx,
+                       const Eigen::Index low, const Eigen::Index high) {
+  if (low < high) {
+    // pi is partitioning index, arr[p] is now at right place
+    const Eigen::Index partition_idx
+        = quick_sort_partition(arr, idx, low, high);
+    // Separately sort elements before partition and after partition
+    if (DoParallel && (high - low >= 400l)) {
+      tbb::parallel_invoke(
+          [&arr, &idx, low, partition_idx]() {
+            quick_sort(arr, idx, low, partition_idx - 1);
+          },
+          [&arr, &idx, high, partition_idx]() {
+            quick_sort(arr, idx, partition_idx + 1, high);
+          });
+    } else {
+      quick_sort<false>(arr, idx, low, partition_idx - 1);
+      quick_sort<false>(arr, idx, partition_idx + 1, high);
+    }
+  }
+}
+
+/**
+ * Runs quick_sort optionally in parallel
```

Comment:
So really I just need a way to both sort the vector and keep track of the original index positions. Let me look at this again and figure out if I can remove it

Reply:
So I looked into how to get rid of this. It's used in [largest_n_elements](https://github.com/stan-dev/stan/pull/3123/files#diff-94e6cc0607001aba25cedde3d8ac0e7bd6ba3bc202556237886fecf2ce385eafR248) to get the largest values of the log likelihood ratio which is then used in `psis_weights` and `multi_pathfinder`. The original indices are used [here](https://github.com/stan-dev/stan/pull/3123/files#diff-94e6cc0607001aba25cedde3d8ac0e7bd6ba3bc202556237886fecf2ce385eafR299) to assign the smoothed likelihood ratio's for the tails.  

If we don't keep track of the original indices then we would instead have to sort the samples in the tail of the matrix of samples by the likelihood ratio so that the indices line up correctly. But that would require copying and a lot of columns around. We could also stop using a matrix to hold the parameters and use a `std::vector<std::vector<double>>` with the lp ratios for each vector store in the inner vector.